### PR TITLE
Scripts and docs for better (docker) user experience 

### PR DESCRIPTION
**Pull request by stolsma**
@jfingerh wanted some more 'helpers' for using this repo with docker and It's always handy to have some helpers ;) so I extended the Makefile with some more docker helpers. This is still "Work In Progress" as I changed a lot in the Makefile and I'm not sure yet what to do with the two script files I added in the volume directory.

What changed? Next to the normal `make run` that just starts the container and stays in the container command line I added `make runp4ovs` that runs the created image as a docker daemon that automatically starts p4-ovs. With `make connect` can then be connected to the running container. Also `make log` can be used to show the output of the container to standard out.  

I Also implement a move to local directory use as running container /tmp volume as I proposed in #34 

The README and other documentation files need to be adjusted to this new proposed way of working, will add that after agreement on the way to go...

Waiting for comments! 👍

### CODE REVIEW

**Comment by stolsma** [comment]
Oeps, I forgot to add the changed Makefile to the commit. Should be there now... Will also do a rebase...

**Comment by jfingerh** [comment]
@stolsma Thanks for the ping, but note that I am still quite new to this code, and new enough to Docker that I am at least not yet the person who should be deciding whether changes like this are desirable to be merged in.

**Comment by stolsma** [comment]
No problem, didn't expect you to decide on merge 😄 Stil working on a change of the readme to show how to use the new docker `make` commands I added. Maybe you can review if the text is clear enough to understand for novices 😄

**Comment by asafravid** [comment]
@stolsma 
Can you please elaborate on the difference between `make volume` and `make connect`?
Much appreciated

File: build/IPDK_Container/docker/README_DOCKER.md
Comment (mestery) [inline build/IPDK_Container/docker/README_DOCKER.md:97]: Since this is a markdown file, it's nice to make example commands wrapped with backticks so users can easily copy and paste them into a terminal. I'd do this everywhere in this file where you have sample commands.

File: build/IPDK_Container/docker/README_DOCKER.md
Comment (mestery) [inline build/IPDK_Container/docker/README_DOCKER.md:107]: Not sure why there is a `sudo` and `sudo less` version here. I'd remove one of them. Also, I'd wrap them in backticks or triple backticks.

File: build/IPDK_Container/docker/README_DOCKER.md
Comment (mestery) [inline build/IPDK_Container/docker/README_DOCKER.md:163]: The [rundemo.sh](https://github.com/ipdk-io/ipdk/blob/main/build/IPDK_Container/scripts/rundemo.sh) script does this automatically using cloud-init. I suggest you just use that rather than having these commands here, as it's much more error prone to have to enter them manually.

**Review by mestery — CHANGES_REQUESTED** [review]
Hi @stolsma this is nice. But I have a few concerns outside of the initial comments:

1. Have you run `make spellcheck` to ensure this passes bash static analysis?
2. I see the approach of an `ipdk` command, and it moves in an interesting direction.
3. The README.md file you have here doesn't seem to actually use any of the new commands and libraries you have in the shell scripts.
4. You would need to update `rundemo.sh` to use your new scripts.
5. You would need to update `README_HOST_INSTALL.md` to reference the new scripts and libraries.
6. I propose we split out `README_HOST_INSTALL.md` a bit and create a separate `README_VAGRANT.md` as well.

Good approach here, but a bit of work remaining, we may want to merge this in pieces in fact.

**Comment by stolsma** [comment]
@mestery yeah agree on all of your remarks. Regarding the README: I was only half ready editing it so tomorrow I will push the 90% version... I wanted to get it on github to have a backup because I needed to work on something else...

**Review by mestery — CHANGES_REQUESTED** [review]
I'll try this out tomorrow, will follow the instructions to see how it works. Nice work, but please don't merge until I've tried it!

**Comment by stolsma** [comment]
> I'll try this out tomorrow, will follow the instructions to see how it works. Nice work, but please don't merge until I've tried it!

I think I'm 85% there so I hope thats enough to start testing but I have 6 hours left tomorrow to get it ready 😊 Happy this time that time zones exist 😎

And don't be afraid that I will merge because I don't have commit rights on this repository 😂

**Comment by stolsma** [comment]
@mestery Just pushed a new update that should bring the total to about 95% ready. The demo for docker should work as described in the README_DOCKER. At least, it works on my laptop 😃 hopefully also at your system...

Have some TODO's (+ your comments that are in the pipeline) left to work on before it can be merged I think. 

I will play with it for another 2 hours and then the weekend will start down here so I hope I will be able to react to any blocking  comments before Monday (depending on the Mrs. 😄) But hopefully everything works as intended!!

Good weekend everybody!

**Comment by mestery** [comment]
Nice work @stolsma! I've pulled this down and I'm following the instructions now to verify this. I may push some commits to this branch for simple things such as fixing shellcheck warnings if you're ok with that:

```
➜  IPDK_Container git:(stolsma-dockercommands) make shellcheck
shellcheck -x `find . -name "*.sh"`

In ./scripts/start.sh line 8:
  source p4ovs_env_setup.sh /root/p4-sde/install
         ^----------------^ SC1091 (info): Not following: p4ovs_env_setup.sh: openBinaryFile: does not exist (No such file or directory)


In ./scripts/start.sh line 26:
  source $HOME/.bashrc
         ^-----------^ SC1091 (info): Not following: ./.bashrc: openBinaryFile: does not exist (No such file or directory)
         ^---^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
  source "$HOME"/.bashrc


In ./scripts/start.sh line 32:
  gnmi-cli $1 $2
           ^-- SC2086 (info): Double quote to prevent globbing and word splitting.
              ^-- SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
  gnmi-cli "$1" "$2"


In ./scripts/start.sh line 36:
  source $HOME/.bashrc
         ^-----------^ SC1091 (info): Not following: ./.bashrc: openBinaryFile: does not exist (No such file or directory)
         ^---^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
  source "$HOME"/.bashrc


In ./scripts/start.sh line 38:
  ${@}
  ^--^ SC2068 (error): Double quote array expansions to avoid re-splitting elements.


In ./scripts/start.sh line 61:
    gnmi_cli $2 $3
             ^-- SC2086 (info): Double quote to prevent globbing and word splitting.
                ^-- SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
    gnmi_cli "$2" "$3"


In ./scripts/start.sh line 65:
    execute ${@}
            ^--^ SC2068 (error): Double quote array expansions to avoid re-splitting elements.

For more information:
  https://www.shellcheck.net/wiki/SC2068 -- Double quote array expansions to ...
  https://www.shellcheck.net/wiki/SC1091 -- Not following: ./.bashrc: openBin...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
make: *** [shellcheck] Error 1
➜  IPDK_Container git:(stolsma-dockercommands)

**Comment by mestery** [comment]
@stolsma Can you check the box to allow maintainers to push commits to your branch? I fixed `ipdk demo` to work and also all the shellcheck errors. I'd like to push my commits to this. Once that is done, I think we're close to merging this! Thanks!

**Comment by mestery** [comment]
BTW: I followed the directions, and with my commits, everything works including the demo of pinging between nested VMs. Very nice work, super clean!

**Comment by stolsma** [comment]
@mestery Nice!!!  I'm still working on `start.sh`, it is not in a shape I wanted it to be but I formed an idea last night which I want to implement Monday! 

Also have a solution for the `ipdk install` process so users don't need to do an `export PATH=` if they are on Ubunto or Centos. And maybe the solution also works on other distributions. 

Will try to find the box to let you push commits!

Yeah, I think I still need two more days to get the PR in 99% ready state and then its up to you and the other maintainers to give it another review and decide if you want to merge it. But looking at your previous comment that wont be a problem I think 😂

**Comment by mestery** [comment]
I'll review this in detail later during my afternoon, but I was able to push my commits to this branch, so thanks for that @stolsma! Nice work and I'm likely to merge this later today, stay tuned.

**Comment by stolsma** [comment]
@mestery I think my changes from my last forced push are being reverted by your new pushes 😞 Could you please check?

**Comment by mestery** [comment]
@stolsma It looks like you collapsed to a single commit which is still there, can you verify? I added my two commits on top.

**Comment by stolsma** [comment]
@mestery It is still there but I found it strange that you added the gnmi-cli function again and reverted my addition of > /dev/null additions...

**Comment by mestery** [comment]
@stolsma Ah, I might have messed up the cherry-pick, I'll have a look later today once my meetings are done and clean it up.

**Comment by mestery** [comment]
@stolsma I force pushed to remove the last shellcheck commit, I didn't need that as you had fixed shellcheck. I added now a single commit. I have a few hours of meetings, then will retest this afternoon and merge at that point, assuming it looks good.

**Pull request merged**
